### ISSUE TITLE
Parent nodes in tree-like recursive structure always appear as child of their children

### ISSUE THREAD

**Issue by gnapse** [emberjs/data#4991]
I think I found a bug in Ember Data, but at the same time I'm very wary of thinking that. However, I don't know what else to make of the behavior I am getting.

Long story short, I created [this repository](https://github.com/gnapse/ember-data-dummy-poc) where the behavior in question can be reproduced. Some details are also given here (mdehoog/ember-data-has-many-query#24) since this was initially thought to be caused by using that ember addon, but the repository I created allows to reproduce the problem without that addon being involved.

In an attempt to briefly describe the problem here for completeness, here it goes:

I'm using Ember Data with a [jsonapi](jsonapi.org/format/) api, and I have a tree-like nested/recursive model `Role`, where each role has many subordinate roles under it in an organization (the children), and also each role has a manager (the parent node).

JsonApi allows to include the URLs to fetch relationships. If the URL given for a certain role, to fetch its subordinates, contains a query param, then for some reason the instances of subordinates obtained when fetching from that URL now include their parent also in the list of their subordinates.

This is perhaps better understood with an example:

```
CEO (id = 1)
  - CTO
    - Software Division Manager
    - Hardware Division Manager
```

If the URL to fetch the CEO's subordinates is `/roles/1/subordinates` then everything works correctly. However, if the URL is `/roles/1/subordinates?depth=1` (regardless of the meaning of this parameter, assume the parameter does not actually change what the api responds) then the roles obtained from that request now include not only their actual subordinates, but they also include their parent node in the list of subordinates.

For instance, in the tree above, after having obtained the subordinates of the CEO, in this case that's the list with only one element, the CTO. If we ask the CTO to give us its subordinates, it'll return a list with three roles: the two roles actually under it in the structure, and the CEO as well.

**Comment by stefanpenner** [comment]
this is due to the default of ember-data models to infer there inverse relationship and keep them in sync.
TL;DR ember-data assumes a graph, and you must opt-into a tree if you want.

```js
DS.Model.extend({
  children: DS.hasMany('node')
})
```

basically assumes the relationship is bidirectional, so a nodes children will have that node in its children.

The following should not:

```js
DS.Model.extend({
  children: DS.hasMany('node', { inverse: null ))
})
```

**Comment by gnapse** [comment]
@stefanpenner I'm sorry, but this does not look like acceptable behavior. If the json-api expresses that the subordinates relationship has two specific roles in it, then why does a third one gets introduced in the mix.

Also the above explanation is not consistent with what's happening. Otherwise how would you explain that when the query param is not present in the relationship link url, the behavior is not present and the data model behaves correctly, as expected?

**Comment by gnapse** [comment]
BTW, I bet this problem has probably not been reported before, because in the absence of the feature asked for in #3700, it is probably uncommon to have people fetching hasMany relationships with URL links having a query param.

I insist that a key characteristic of this behavior, and part of why I'm almost convinced is some sort of bug, is that **it occurs only when the relationship links have the query param**, even if the data returned from the backend does not change with the presence or not of query param.

**Comment by gnapse** [comment]
Declaring the inverse relationship fixes the problem. However I still don't understand why this is not necessary when the link has no query param, but it becomes necessary when it has some.

**Comment by runspired** [comment]
@gnapse @stefanpenner after looking into this, I created a simpler demo in twiddle that shows this is not a bug if inverse relationships are added (when added to the POC the POC works as well): https://ember-twiddle.com/734a5637cadae3bd9e2f6fabd9b9421b?openFiles=adapters.application.js%2C

I suspect that the real bug is this one: https://github.com/emberjs/data/issues/4712

Basically, without inverses being explicitly defined, the "auto-inverse-discovery" is causing manager to also be added to subordinates.  With the link+data bug, this means that the link will never be used and instead a url will be generated to fetch each subordinate, leaving the manager in the array incorrectly.
